packages/plugin-ext/src/main/browser/tabs/tabs-main.ts

```diff
@@ -15,28 +15,312 @@
+            this.proxy.$acceptTabGroupUpdate(tabInfo.group);
+        }
+        if (!this.tabDtosEqual(oldTabDto, newTabDto)) {
+            tabInfo.group.tabs.splice(tabInfo.tabIndex, 1, newTabDto);
+            this.proxy.$acceptTabOperation({
+                kind: TabModelOperationKind.TAB_UPDATE,
+                index: tabInfo.tabIndex,
+                tabDto: newTabDto,
+                groupId: tabInfo.group.groupId
+            });
+        }
+    }
+
+    private onTabClosed(tabInfo: TabInfo, title: Title<Widget>): void {
+        tabInfo.group.tabs.splice(tabInfo.tabIndex, 1);
+        this.tabInfoLookup.delete(title);
+        this.proxy.$acceptTabOperation({
+            kind: TabModelOperationKind.TAB_CLOSE,
+            index: tabInfo.tabIndex,
+            tabDto: this.createTabDto(title, tabInfo.group.groupId),
+            groupId: tabInfo.group.groupId
+        });
+    }
+
+    private onTabMoved(tabBar: TabBar<Widget>, args: TabBar.ITabMovedArgs<Widget>): void {
+        const tabInfo = this.getOrRebuildModel(this.tabInfoLookup, args.title)!;
+        tabInfo.tabIndex = args.toIndex;
+        const tabDto = this.createTabDto(args.title, tabInfo.group.groupId);
+        tabInfo.group.tabs.splice(args.fromIndex, 1);
+        tabInfo.group.tabs.splice(args.toIndex, 1, tabDto);
+        this.proxy.$acceptTabOperation({
+            kind: TabModelOperationKind.TAB_MOVE,
+            index: args.toIndex,
+            tabDto,
+            groupId: tabInfo.group.groupId,
+            oldIndex: args.fromIndex
+        });
+    }
+
+    private onTabGroupCreated(tabBar: TabBar<Widget>): void {
+        this.tabGroupChanged = true;
+    }
+
+    private onTabGroupClosed(tabBar: TabBar<Widget>): void {
+        this.tabGroupChanged = true;
+    }
+    // #endregion
 
     // #region Messages received from Ext Host
     $moveTab(tabId: string, index: number, viewColumn: number, preserveFocus?: boolean): void {
         return;
     }
 
     async $closeTab(tabIds: string[], preserveFocus?: boolean): Promise<boolean> {
-        return false;
+        for (const tabId of tabIds) {
+            const widget = this.applicationShell.getWidgetById(tabId.substring(tabId.indexOf('~') + 1));
+            if (!widget) {
+                continue;
+            }
+            if (widget.title === this.applicationShell.mainPanel.currentTitle) {
+                this.applicationShell.activateNextTabInTabBar();
+            }
+            widget.dispose();
+
+        }
+        return true;
     }
 
     async $closeGroup(groupIds: number[], preserveFocus?: boolean): Promise<boolean> {
-        return false;
+        for (const groupId of groupIds) {
+            const tabBar = Array.from(this.tabGroupModel.entries()).find(([, tabGroup]) => tabGroup.groupId === groupId)?.[0];
+            if (tabBar) {
+                tabBar.titles.forEach(title => title.owner.dispose());
+            }
+        }
+        return true;
```

Comment:
Bug: The current code doesn't close the active widget if there is more than one widget in a tab group.

```suggestion
        const allWidgets: Widget[] = [];
        for (const groupId of groupIds) {
            const tabBar = Array.from(this.tabGroupModel.entries()).find(([, tabGroup]) => tabGroup.groupId === groupId)?.[0];
            if (tabBar) {
                allWidgets.push(...tabBar.titles.map(title => title.owner));
            }
        }
        await this.applicationShell.closeMany(allWidgets);
        return true;
```